### PR TITLE
SI-7814 Avoid init cycle between Predef, `package`, ScalaRuntime

### PR DESCRIPTION
**Pull request by retronym**
Not every application will force these in a single thread; we
have to do our best to avoid cycles between them.

The enclosed test was failing every time before the change.

This commit breaks the cycle by avoiding computing `tupleNames`
in the constructor of `ScalaRuntime`. The new version has the
added benefit of including specialized tuple subclasses, which
is verified with a unit test for `isTuple`.

Are there more of these lurking? It seems likely. I'm more than
a little concerned about the way the `ControlThrowable` fires up
`scala.SystemProperties` to check whether or not to suppress
stack traces; there is already an ugly hack in place:

```
object NoStackTrace {
  final def noSuppression = _noSuppression

  // two-stage init to make checkinit happy,
  // since sys.SystemProperties.noTraceSupression.value
  // calls back into NoStackTrace.noSuppression
  final private var _noSuppression = false
  _noSuppression = sys.SystemProperties.noTraceSupression.value
}
```

Review by @paulp

### CODE REVIEW

**Comment by paulp** [comment]
Some rationale for the joint existence of these is long overdue as well. Ask anyone why some particular thing is in the scala package object instead of Predef, or vice versa, and see what sort of answers you receive.

NoStackTrace is one example of the general difficulty of configuring things transparently. If you know a better way to have some flags which influence global behaviors, it will probably sound great to me. It would be nice to have a lot more of them, but I've stayed away due to the obvious suboptimality of using system properties.

File: src/library/scala/runtime/ScalaRunTime.scala [hunk 1]
```diff
@@ -75,33 +67,37 @@ object ScalaRunTime {
...
+      case x: Array[Long]    => x(idx) = value.asInstanceOf[Long]
+      case x: Array[Float]   => x(idx) = value.asInstanceOf[Float]
+      case x: Array[Char]    => x(idx) = value.asInstanceOf[Char]
+      case x: Array[Byte]    => x(idx) = value.asInstanceOf[Byte]
+      case x: Array[Short]   => x(idx) = value.asInstanceOf[Short]
+      case x: Array[Boolean] => x(idx) = value.asInstanceOf[Boolean]
+      case x: Array[Unit]    => x(idx) = value.asInstanceOf[Unit]
+      case null => throw new NullPointerException
+    }
   }
```
Comment (paulp) [inline src/library/scala/runtime/ScalaRunTime.scala:101]: Gratuitous reformatting?

File: test/instrumented/library/scala/runtime/BoxesRunTime.java [hunk 2]
```diff
@@ -244,7 +244,7 @@ private static boolean equalsNumChar(java.lang.Number xn, java.lang.Character yc
      *  as yet have not.
      *
      *  Note: Among primitives, Float.NaN != Float.NaN, but the boxed
-     *  verisons are equal.  This still needs reconciliation.
+     *  versions are equal.  This still needs reconciliation.
```
Comment (paulp) [inline test/instrumented/library/scala/runtime/BoxesRunTime.java:247]: A good verison stew is delicious.

File: src/library/scala/runtime/ScalaRunTime.scala (hunk 1 above)
Comment (paulp) [inline src/library/scala/runtime/ScalaRunTime.scala:101]: Oh, I hadn't seen the second commit message yet.

Comment (paulp) [inline src/library/scala/runtime/ScalaRunTime.scala:101]: I feel like github completely blows it on the presentation of multiple commits.

**Comment by retronym** [comment]
I don't so much mind hitting j.l.System.getProperty, it is the calls into our wrappers that look more prone to take part in a deadlock.

**Comment by retronym** [comment]
I'd say that `scala.package` is just for aliases. Predef avoids, usefully, exposing `scala.println("!")`

**Comment by paulp** [comment]
LGTM

**Pull request merged**